File: sdk/swift/Sources/NaatreCore/Operation.swift
import Foundation

public enum OperationKind: String, Codable, Sendable {
    case query
    case mutation
    case subscription
}

public struct PersistedReference: Codable, Equatable, Sendable {
    public let algorithm: String
    public let canonicalVersion: String
    public let digest: String

    public init(algorithm: String = "sha-256", canonicalVersion: String = "c14n-1", digest: String) throws {
        guard algorithm == "sha-256", canonicalVersion == "c14n-1", digest.count == 64,
              digest.utf8.allSatisfy({ ($0 >= 48 && $0 <= 57) || ($0 >= 97 && $0 <= 102) }) else {
            throw ClientFailure(.protocolInvalid)
        }
        self.algorithm = algorithm
        self.canonicalVersion = canonicalVersion
        self.digest = digest
    }
}

public struct Operation<Variables: Encodable & Sendable, Result: Codable & Sendable>: Sendable {
    public let name: String
    public let kind: OperationKind
    public let persisted: PersistedReference
    public let variables: Variables

    public init(name: String, kind: OperationKind, persisted: PersistedReference, variables: Variables) throws {
        guard name.range(of: #"^[A-Za-z_][A-Za-z0-9_]{0,127}$"#, options: .regularExpression) != nil else {
            throw ClientFailure(.protocolInvalid)
        }
        self.name = name
        self.kind = kind
        self.persisted = persisted
        self.variables = variables
    }

    public func canonicalRequest() throws -> Data {
        let request = Request(version: "1", operation: name, persisted: persisted, variables: variables)
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys, .withoutEscapingSlashes]
        return try encoder.encode(request)
    }

    public func decodeResult(_ data: Data) throws -> OperationResult<Result> {
        do { return try JSONDecoder().decode(OperationResult<Result>.self, from: data) }
        catch let failure as ClientFailure { throw failure }
        catch { throw ClientFailure(.resultInvalid) }
    }

    private struct Request: Encodable {
        let version: String
        let operation: String
        let persisted: PersistedReference
        let variables: Variables
    }
}

public struct ManifestOperation: Codable, Equatable, Sendable {
    public let name: String
    public let kind: OperationKind
    public let persisted: PersistedReference

    public init(name: String, kind: OperationKind, persisted: PersistedReference) {
        self.name = name
        self.kind = kind
        self.persisted = persisted
    }
}

public struct Manifest: Codable, Equatable, Sendable {
    public let profile: String
    public let version: String
    public let protocolVersion: String
    public let canonicalVersion: String
    public let operations: [ManifestOperation]

    public init(profile: String = "sdk.swift.core-1", operations: [ManifestOperation]) throws {
        guard !operations.isEmpty, Set(operations.map(\.name)).count == operations.count else {
            throw ClientFailure(.protocolInvalid)
        }
        self.profile = profile
        version = "1"
        protocolVersion = "1"
        canonicalVersion = "c14n-1"
        self.operations = operations.sorted { $0.name < $1.name }
    }
}

File: sdk/typescript/runtime/operation.mjs
import { fail } from "./error.mjs";
import { canonicalStringify, hasOwn, parseJSON, safeObject } from "./json.mjs";
import { decodeOperationResult } from "./result.mjs";
import { createScalarCodecs, encodeScalar } from "./scalars.mjs";

const operationNames = /^(?:[A-Za-z_][A-Za-z0-9_]{0,127})$/u;
const digest = /^[0-9a-f]{64}$/u;

export function createOperation(definition, variables, decodeData, options = {}) {
  definition = validateDefinition(definition);
  if (!isRecord(variables) || Object.getOwnPropertySymbols(variables).length !== 0) fail("CLIENT_VARIABLES_INVALID");
  const codecs = options.codecs ?? createScalarCodecs();
  const encodedVariables = encodeVariables(definition.variables, variables, codecs);
  const request = Object.freeze({
    version: "1",
    operation: definition.name,
    persisted: definition.persisted,
    variables: Object.freeze(encodedVariables),
  });
  const canonical = canonicalStringify(request);
  return Object.freeze({
    kind: definition.kind,
    request,
    canonicalRequest: () => canonical,
    decodeResult: (input) => decodeOperationResult(input, decodeData),
  });
}

function encodeVariables(variableDefinitions, variables, codecs) {
  const allowed = new Map(variableDefinitions.map((entry) => [entry.name, entry]));
  const result = Object.create(null);
  const descriptors = Object.getOwnPropertyDescriptors(variables);
  for (const key of Object.keys(descriptors)) {
    const descriptor = descriptors[key];
    const entry = allowed.get(key);
    if (!descriptor?.enumerable || !("value" in descriptor) || !entry || descriptor.value === undefined) fail("CLIENT_VARIABLES_INVALID");
    result[key] = encodeVariable(entry, descriptor.value, codecs);
  }
  for (const entry of allowed.values()) {
    if (entry.required && !hasOwn(result, entry.name)) fail("CLIENT_VARIABLES_INVALID");
  }
  return result;
}

function encodeVariable(definition, value, codecs) {
  if (value !== null) return encodeScalar(codecs, definition.type, value);
  if (!definition.nullable) fail("CLIENT_VARIABLES_INVALID");
  return null;
}

export function loadManifest(input) {
  const manifest = typeof input === "string" ? parseJSON(input, 1 << 20) : safeObject(input, "CLIENT_MANIFEST_INVALID");
  if (!hasExactKeys(manifest, ["canonicalVersion", "operations", "profile", "protocolVersion", "version"])) fail("CLIENT_MANIFEST_INVALID");
  if (manifest?.profile !== "sdk.typescript.core-1" || manifest.version !== "1" || manifest.protocolVersion !== "1" || manifest.canonicalVersion !== "c14n-1" || !Array.isArray(manifest.operations) || manifest.operations.length === 0) fail("CLIENT_MANIFEST_INVALID");
  const operations = Object.create(null);
  for (const operation of manifest.operations) {
    if (!isRecord(operation) || !hasExactKeys(operation, ["kind", "name", "persisted"])) fail("CLIENT_MANIFEST_INVALID");
    const validated = validateDefinition({ ...operation, variables: [] });
    if (hasOwn(operations, operation.name)) fail("CLIENT_MANIFEST_INVALID");
    operations[operation.name] = Object.freeze({ name: validated.name, kind: validated.kind, persisted: validated.persisted });
  }
  return Object.freeze(operations);
}

function validateDefinition(definition) {
  const value = safeObject(definition, "CLIENT_OPERATION_INVALID");
  if (!hasExactKeys(value, ["kind", "name", "persisted", "variables"]) || !isRecord(value.persisted) || !hasExactKeys(value.persisted, ["algorithm", "canonicalVersion", "digest"]) || !operationNames.test(value.name ?? "") || !["query", "mutation", "subscription"].includes(value.kind) || value.persisted.algorithm !== "sha-256" || value.persisted.canonicalVersion !== "c14n-1" || !digest.test(value.persisted.digest ?? "") || !Array.isArray(value.variables)) fail("CLIENT_OPERATION_INVALID");
  const names = new Set();
  for (const variable of value.variables) {
    if (!isRecord(variable) || !hasExactKeys(variable, ["name", "nullable", "required", "type"]) || !operationNames.test(variable?.name ?? "") || !operationNames.test(variable?.type ?? "") || typeof variable.required !== "boolean" || typeof variable.nullable !== "boolean" || names.has(variable.name)) fail("CLIENT_OPERATION_INVALID");
    names.add(variable.name);
  }
  return Object.freeze({
    name: value.name,
    kind: value.kind,
    persisted: Object.freeze({ algorithm: value.persisted.algorithm, canonicalVersion: value.persisted.canonicalVersion, digest: value.persisted.digest }),
    variables: Object.freeze(value.variables.map((variable) => Object.freeze({ name: variable.name, type: variable.type, required: variable.required, nullable: variable.nullable }))),
  });
}

function isRecord(value) {
  if (value === null || typeof value !== "object" || Array.isArray(value)) return false;
  const prototype = Object.getPrototypeOf(value);
  return prototype === Object.prototype || prototype === null;
}

function hasExactKeys(value, expected) {
  const keys = Object.keys(value).sort();
  return keys.length === expected.length && keys.every((key, index) => key === expected[index]);
}
